cfpb/cfgov-refresh
# Df 1860 validate forms

Created two new CF jQuery plugins to support validating forms and displaying a message when that validation fails.

## Additions

- Adds validate library to handle validation logic.
- Adds handlebars library for JS templating.
- Adds cf_formValidator for validating form inputs for multiple types and returning an object containing the valid and invalid inputs.
- Adds cf_notifier plugin for creating a notification using a template and data objects that are customizable.
- Adds cf-notifications stylesheet to be ported to CF in the future

## Removals

- Removes alert.js because not all messages are alerts, the term "notification" covers more of our bases for success and warning messages in the future.
- Removes the alert macro so that we can insert messages as needed and remove them when we're done.

## Changes

- Updated cf-enhancements to include warning states for notifications and form input highlights

## Testing

- Checkout branch locally
- navigate to `/blog/` or `/newsroom/` and use the subscribe form
- navigate to `/the-bureau/leadership-calendar/` and use the filtering form at the top (this will be updated, this is just for ease of testing the check groups)

## Review

- @everyone

## Preview
![screen shot 2015-05-13 at 5 12 47 pm](https://cloud.githubusercontent.com/assets/1280430/7621259/6462962a-f993-11e4-9cca-d80efac7940d.png)

![screen shot 2015-05-13 at 5 12 57 pm](https://cloud.githubusercontent.com/assets/1280430/7621261/646ee060-f993-11e4-9cfb-e992b357c15e.png)

![screen shot 2015-05-13 at 5 13 17 pm](https://cloud.githubusercontent.com/assets/1280430/7621260/6466ddf2-f993-11e4-8beb-649358a1fc27.png)

[Preview this PR without the whitespace changes](?w=0)

## Notes

- There's still a lot that could be done, but the plugin is in a pretty solid state. Some expansion will probably occur when adding the subscribe ajax feature.
- Before merging down the origin, I wanted to get feedback and make updates first.
- Will add tests and update the changelog after merging down origin and ping everyone again.
- Using the namespace pattern for jQuery libraries to avoid breaking everything if an unknown method is passed and to ease up on passing data around from function to function.
- Odd space between border and outline in Safari is a known bug.
